protocols/autonat/src/protocol.rs

```diff
@@ -0,0 +1,325 @@
+            peer:
+                Some(structs_proto::message::PeerInfo {
+                    id: Some(peer_id),
+                    addrs,
+                }),
+        }) = msg.dial
+        {
+            (peer_id, addrs)
+        } else {
+            log::debug!("Received malformed dial message.");
+            return Err(io::Error::new(
+                io::ErrorKind::InvalidData,
+                "invalid dial message",
+            ));
+        };
+
+        let peer_id = {
+            PeerId::try_from(peer_id)
+                .map_err(|_| io::Error::new(io::ErrorKind::InvalidData, "invalid peer id"))?
+        };
+        let addrs = {
+            let mut maddrs = vec![];
+            for addr in addrs.into_iter() {
+                let maddr = Multiaddr::try_from(addr)
+                    .map_err(|err| io::Error::new(io::ErrorKind::InvalidData, err))?;
+                maddrs.push(maddr);
+            }
+            maddrs
+        };
+        Ok(Self {
+            peer_id,
+            addresses: addrs,
+        })
+    }
+
+    pub fn into_bytes(self) -> Vec<u8> {
+        let peer_id = self.peer_id.to_bytes();
+        let addrs = self
+            .addresses
+            .into_iter()
+            .map(|addr| addr.to_vec())
+            .collect();
+
+        let msg = structs_proto::Message {
+            r#type: Some(structs_proto::message::MessageType::Dial as _),
+            dial: Some(structs_proto::message::Dial {
+                peer: Some(structs_proto::message::PeerInfo {
+                    id: Some(peer_id),
+                    addrs,
+                }),
+            }),
+            dial_response: None,
+        };
+
+        let mut bytes = Vec::with_capacity(msg.encoded_len());
+        msg.encode(&mut bytes)
+            .expect("Vec<u8> provides capacity as needed");
+        bytes
+    }
+}
+
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub enum ResponseError {
+    DialError,
+    DialRefused,
+    BadRequest,
+    InternalError,
+}
+
+impl From<ResponseError> for i32 {
+    fn from(t: ResponseError) -> Self {
+        match t {
+            ResponseError::DialError => 100,
+            ResponseError::DialRefused => 101,
+            ResponseError::BadRequest => 200,
+            ResponseError::InternalError => 300,
+        }
+    }
+}
+
+impl TryFrom<i32> for ResponseError {
+    type Error = io::Error;
+
+    fn try_from(value: i32) -> Result<Self, Self::Error> {
+        match value {
+            100 => Ok(ResponseError::DialError),
+            101 => Ok(ResponseError::DialRefused),
+            200 => Ok(ResponseError::BadRequest),
+            300 => Ok(ResponseError::InternalError),
+            _ => {
+                log::debug!("Received response with invalid status code.");
+                Err(io::Error::new(
+                    io::ErrorKind::InvalidData,
+                    "invalid response error type",
+                ))
+            }
+        }
+    }
+}
```

Comment:
The generated `ResponseStatus` should implement `TryFrom<i32>` right? If so, how about leveraging that implementation, thus not having two sources of truth on the number values, and manually implement `ResponseStatus` -> `ResponseError`?

```suggestion
impl TryFrom<ResponseStatus> for ResponseError {
    type Error = io::Error;

    fn try_from(value: ResponseStatus) -> Result<Self, Self::Error> {
        match value {
            ResponseStatus::DialError => Ok(ResponseError::DialError),
            // ...
        }
    }
}
```

Reply:
> The generated ResponseStatus should implement TryFrom<i32> right?

Is that done by `prost_build`, or why should it be implemented? If so, could you link me any docs where that's described?
I tried to follow your proposal but I get the error that `TryFrom<i32>` is not implemented for `ResponseStatus`.

Reply:
Yes, it is generated by `prost_build` for `ResponseStatus`. Not sure whether it is documented anywhere.

See https://github.com/elenaf9/rust-libp2p/pull/3.